### PR TITLE
[traverse] reverse meaning of isolation in difference

### LINKED ISSUE #869: Sym difference may generate polygon with disconnected interior
Please see this example:
https://wandbox.org/permlink/Ib8ZrLmUlRI8UZEU

````C++
#include <iostream>

#include <boost/geometry.hpp>
#include <boost/geometry/geometries/point_xy.hpp>
#include <boost/geometry/geometries/polygon.hpp>
#include <boost/geometry/geometries/multi_polygon.hpp>

using point_t = boost::geometry::model::d2::point_xy<double>;
using polygon_t = boost::geometry::model::polygon<point_t>;
using ring_t = boost::geometry::model::ring<point_t>;
using multi_polygon_t = boost::geometry::model::multi_polygon<polygon_t>;


int main()
{
    multi_polygon_t a;
    multi_polygon_t b;
    multi_polygon_t c;
    
    boost::geometry::read_wkt("MULTIPOLYGON(((70 50,70 10,10 10,10 70,50 70,50 90,70 90,70 70,90 70,90 50,70 50)))", a);
    boost::geometry::read_wkt("MULTIPOLYGON(((50 50,50 70,70 70,70 50,50 50)))", b);
    boost::geometry::read_wkt("MULTIPOLYGON(((50 50,50 30,30 30,30 50,50 50)))", c);
    
    multi_polygon_t result;
    boost::geometry::sym_difference(b, c, result);
    std::cout << boost::geometry::wkt(result) << std::endl;
    
    multi_polygon_t result_2;
    boost::geometry::sym_difference(result, a, result_2);
    
    std::string message;
    if(!boost::geometry::is_valid(result_2, message))
        std::cout << "Geometry is not valid: " << message << std::endl;
    
    std::cout << boost::geometry::wkt(result_2) << std::endl;
    return 0;
}
````

Expected output is a valid polygon, output is: 
Geometry is not valid: Geometry has disconnected interior
MULTIPOLYGON(((50 70,50 90,70 90,70 70,90 70,90 50,70 50,70 10,10 10,10 70,50 70),(50 70,50 50,70 50,70 70,50 70),(50 50,30 50,30 30,50 30,50 50)))

### PR DESCRIPTION
**Pull request by barendgehrels**
This fixes #869 , the recently found problem in difference and interconnected interiors, and should fix similar cases too (though they were not yet in the test suite).

The root cause was that isolated interior rings were detected but in case of difference, it's meaning should be reversed.

### CODE REVIEW

**Comment by barendgehrels** [comment]
Note that I get an unrelated error
```
clang-linux.compile.c++.without-pch detail/bin/algorithms_visit.test/clang-linux-6.0.0/debug/visit.o
In file included from detail/visit.cpp:18:
../boost/geometry/geometries/adapted/std_any.hpp:42:16: error: no template named 'any_cast' in namespace 'std'; did you mean simply 'any_cast'?
        return std::any_cast<T>(any_ptr);
               ^~~~~~~~~~~~~
               any_cast
../boost/any.hpp:249:17: note: 'any_cast' declared here
    ValueType * any_cast(any * operand) BOOST_NOEXCEPT
```
but maybe it's fixed in the meantime

File: include/boost/geometry/algorithms/detail/overlay/traversal_switch_detector.hpp [hunk 1]
```diff
@@ -388,13 +387,36 @@ struct traversal_switch_detector
...
+                    // It is assigned to isolated if it's property is "Yes",
+                    // (one connected interior, or chained).
+                    // "Multiple" doesn't count for isolation,
+                    // neither for intersection, neither for difference.
                     region_properties const& prop = mit->second;
-                    op.enriched.isolated = prop.isolated == isolation_yes;
+                    op.enriched.isolated
+                            = reverseMeaningInOp
+                            ? prop.isolated == isolation_no
+                            : prop.isolated == isolation_yes;
```
Comment (barendgehrels) [inline include/boost/geometry/algorithms/detail/overlay/traversal_switch_detector.hpp:419]: this is the fix

**Comment by barendgehrels** [comment]
Also I get, unrelated:
```
In file included from ../boost/geometry/strategies/convex_hull/cartesian.hpp:16:
../boost/geometry/strategy/cartesian/side_robust.hpp:53:20: error: non-constant-expression cannot be narrowed from type 'typename coordinate_type<point_xy<int, cartesian> >::type' (aka 'int') to 'double' in initializer list [-Wc++11-narrowing]
        vec2d pa { get<0>(p1), get<1>(p1) };
```

**Comment by awulkiew** [comment]
@barendgehrels I don't know what's heppening with the first one because the code is behind `#ifndef BOOST_NO_CXX17_HDR_ANY`. Maybe there is something wrong with clang-6 and we'd need a workaround. Are you setting `-std`?

Regarding the second one. This was already fixed: https://github.com/boostorg/geometry/pull/878

**Comment by barendgehrels** [comment]
> @barendgehrels I don't know what's heppening with the first one because the code is behind `#ifndef BOOST_NO_CXX17_HDR_ANY`. Maybe there is something wrong with clang-6 and we'd need a workaround. Are you setting `-std`?
> 
> Regarding the second one. This was already fixed: #878

I'm just calling `b2 --toolset=clang` and my clang version is `clang version 5.0.2 (tags/RELEASE_502/final)`

**Comment by vissarion** [comment]
> > @barendgehrels I don't know what's heppening with the first one because the code is behind `#ifndef BOOST_NO_CXX17_HDR_ANY`. Maybe there is something wrong with clang-6 and we'd need a workaround. Are you setting `-std`?
> > Regarding the second one. This was already fixed: #878
> 
> I'm just calling `b2 --toolset=clang` and my clang version is `clang version 6.0.0-1ubuntu2 (tags/RELEASE_600/final)`
> (I've also `clang version 5.0.2 (tags/RELEASE_502/final)` but that gives me other errors, and more)

I am also getting something similar with `gcc`

```
gcc.compile.c++ ../../bin.v2/libs/geometry/test/algorithms/detail/algorithms_visit.test/gcc-5~c++14/debug/threading-multi/visibility-hidden/visit.o
In file included from test/algorithms/detail/visit.cpp:18:0:
../../boost/geometry/geometries/adapted/std_any.hpp:19:15: fatal error: any: No such file or directory
compilation terminated.
```

**Comment by awulkiew** [comment]
@barendgehrels @vissarion AFAIU C++17 headers support was added to Boost.Config in 1.76.
Which version of Boost.Config do you have?
When did you last update the whole Boost (all of the modules)?

**Comment by kleunen** [comment]
I have the following test case which was correct under the under the old version, but seems to be incorrect, after this patch:
https://wandbox.org/permlink/2zBpf9zFb1Uepu7n

**Comment by barendgehrels** [comment]
> 
> 
> @barendgehrels @vissarion AFAIU C++17 headers support was added to Boost.Config in 1.76.
> Which version of Boost.Config do you have?
> When did you last update the whole Boost (all of the modules)?

@awulkiew ah, of course, sorry. Will update soon, it's long ago indeed.

**Comment by barendgehrels** [comment]
> 
> 
> I have the following test case which was correct under the under the old version, but seems to be incorrect, after this patch:
> https://wandbox.org/permlink/2zBpf9zFb1Uepu7n

Thanks! I will look at it soon

**Comment by awulkiew** [comment]
LGTM. Is this a bugfix? Would you like to release it with 1.77 if possible? Could we merge https://github.com/boostorg/geometry/pull/886 before this PR in case you wanted to postpone this one to 1.78?

**Review by awulkiew — APPROVED** [review]

**Comment by barendgehrels** [comment]
> LGTM. Is this a bugfix? Would you like to release it with 1.77 if possible? Could we merge #886 before this PR in case you wanted to postpone this one to 1.78?

Wouter found a problem in another case, I’ve still to check that. Therefore it’s better to postpone it to 1.78

**Comment by barendgehrels** [comment]
> I have the following test case which was correct under the under the old version, but seems to be incorrect, after this patch:
> https://wandbox.org/permlink/2zBpf9zFb1Uepu7n

I can reproduce it and I've a reduced version for it. I will create a separate issue for this, to add this testcase, and will try to fix it (large input... it might take a while).
It's a combination of complex input multipolygon - while the solved case is simple. So alternatively we can merge this already and (try to) fix it afterwards

**Comment by kleunen** [comment]
> Wouter found a problem in another case, I’ve still to check that. Therefore it’s better to postpone it to 1.78

yes, there is no need to rush to a fix. In many cases actually the sym_difference works well. I sort the input polygons now by size and if I do a sym_difference in this order. The output is fine. So I really think it is a corner case and you don't want to break functionality working well for other users.

**Comment by barendgehrels** [comment]
The regression in the two cases referred by #888 should be fixed now

File: include/boost/geometry/algorithms/detail/overlay/traversal_switch_detector.hpp (hunk 1 above)
Comment (barendgehrels) [reply include/boost/geometry/algorithms/detail/overlay/traversal_switch_detector.hpp:419]: The first version did have `prop.isolated == isolation_no` but that is the default, we should not revert all of them into yes.

But it now does not look symmetrical, theoretically there should be cases where reversal should be necessary - but we don't have them currently in our test set.

File: include/boost/geometry/algorithms/detail/overlay/traversal_switch_detector.hpp [hunk 2]
```diff
@@ -565,10 +591,83 @@ struct traversal_switch_detector
         }
     }
 
+#if defined(BOOST_GEOMETRY_DEBUG_TRAVERSAL_SWITCH_DETECTOR)
+    void debug_show_results()
```
Comment (barendgehrels) [inline include/boost/geometry/algorithms/detail/overlay/traversal_switch_detector.hpp:595]: this shows quite a neat list of results and I would like to keep it.

For example
```
BEGIN SWITCH DETECTOR (region_ids and isolation) REVERSE_2
 ADD (s:0, m:0) TO REGION 1
 ADD (s:1, m:0) TO REGION 2
 ADD (s:1, m:1) TO REGION 3
END SWITCH DETECTOR

REGION 1 multiple [turns 0 1 2] regions { 2 : via [ 0 1 2 ] }
REGION 2 yes [turns 0 1 2 3] regions { 1 : via [ 0 1 2 ] } { 3 : via [ 3 ] }
REGION 3 yes [turns 3] regions { 2 : via [ 3 ] }

II 0 (50, 70) ->  [0/1] (1 false) / (2 false)
II 1 (70, 70) ->  [0/1] (1 false) / (2 false)
II 2 (70, 50) ->  [0/1] (1 false) / (2 false)
II 3 (50, 50) ->  [1/1] (2 true) / (3 true)
Analyze II from turn 0 (50, 70) region 1 -> to turn 1 (70, 70) region 2
Analyze II from turn 0 (50, 70) region 2 -> to turn 3 (50, 50) region 2
Analyze II from turn 3 (50, 50) region 2 -> to turn 2 (70, 50) region 1
Analyze II from turn 1 (70, 70) region 1 -> to turn 2 (70, 50) region 2
```

File: include/boost/geometry/algorithms/detail/overlay/traversal_switch_detector.hpp (hunk 1 above)
Comment (kleunen) [reply include/boost/geometry/algorithms/detail/overlay/traversal_switch_detector.hpp:419]: Can you reverse one of the test cases? To make sure the other option also gets triggered by test suite?

Comment (barendgehrels) [reply include/boost/geometry/algorithms/detail/overlay/traversal_switch_detector.hpp:419]: We test all our difference operations in two ways, so both `a - b` and `b - a` . If that is what you mean. 

Or you probably mean (what I call) inverse them, so make a large polygon around a polygon and make the polygon itself an interior of that. Yes, I could do that, I will try to construct a case where this should happen. I now know where I’m looking for. I will do that, but not today.

Thanks.

**Pull request merged**